### ISSUE TITLE
Styling a component with multiple possible decorations

### ISSUE THREAD

**Issue by kgoggin** [css-modules/css-modules#85]
I love the idea of CSS Modules and have started playing around with the concept. I'm looking for some advice on how to handle styling a component with multiple independent groups of decorators, any of which could be applied together.

For example, let's say I'm styling a button component. It can have one of three "type" decorators applied: primary, secondary, or tertiary (our style guide has different use cases for each type). Additionally, it can also have one of three "size" decorators applied: large, medium, or small. Both sets of decorators are independent of each other, so there are 9 possible configurations for this button.

Before CSS Modules, I would have simply made a base `.button` class, as well as type classes and size classes, and applied each as needed. But, if I'm understanding Glen's introductory post correctly, the "right" way to do this with CSS modules would be to create a `.common` class for shared styling, then a class for each decorator (6 total), then classes for each potential combination of decorators that compose from `.common`, a type class, and a size class (9 total). Then, in my component code, I'm using a `type` and `size` prop passed into my component to build my final class name (`.primaryMedium`, `.tertiaryLarge`, etc.). 

That *feels* kind of wrong since each decorator group is independent - I feel like I'm writing a lot more CSS to make it work "The CSS Modules Way". I also worry that, in the future, I'd want to add some other decorator, like `.isSubmitting` or something, and my CSS continues to grow exponentially. But, I can also see the draw of only ever worrying about applying one class name to my component.

One idea I had was to handle they sizing by building different Components for each size (`ButtonLarge`, `ButtonMedium`, `ButtonSmall`) since the size would theoretically never change once rendered. But I'm not sure I love that either because it the size really is a presentation detail, which makes sense to live in the CSS alone.

So, I'm just looking for others' thoughts on this. Am I understanding the overall philosophy correctly? Anyone else bumped into a situation like this? Seems like any implementation of Bootstrap, for example, would need to deal with this pretty quickly.

**Comment by ojame** [comment]
If your `type` and `size` prop values directly correlate to classes, you could do something as simple as:

```
const buttonClass = classnames({
   [styles[this.props.type]]: true,
   [styles[this.props.size]]: true,
});
```

**Comment by nkbt** [comment]
That would work well too
```js
<button className={`${styles[this.props.type]} ${styles[this.props.size]}`}
```

**Comment by geelen** [comment]
Yeah look if they're _really_ independent then simply toggle through the `type` and `size` classes. One of the good things about CSS Modules is that no matter what all your code is in the one file, so if you want to have 9 classes or 6 or whatever, it's ok. Do whatever makes sense in your context.

The reason I favour _explicitness_ over conciseness is from two points. The first is that usually the things aren't truly independent — for some reason the "small" variant never appears the "primary" variant or something. The second is that maybe having to list every combination is a good constraint for the project — it helps stop an explosion of visual elements across the site. If you have 9 buttons now, do you want one line of code to make a 10th button or one extra size class to make 18 total button types?

So I'd probably do this:

```css
.base { /* truly common stuff */ }

/* single variations */
.small { composes: base; /* ... */ }
.large { composes: base; /* ... */ }
.primary { composes: base; /* ... */ }
.danger { composes: base; /* ... */ }

/* variation combinations */
.small-primary { composes: small primary; }
.small-danger { composes: small danger; }
.large-primary { composes: large primary; }
.large-danger { composes: large danger; }
```

For me, having to enumerate all the combinations isn't such a bad thing, as long as you pass those names around your app. So, do this:

```jsx
let ButtonComponent = ({variant, children} => (
  <button className={variant}>{props.children}
</button>)

return <ButtonComponent variant="large-danger">Delete project!</ButtonComponent>
```

not this:

```jsx
let ButtonComponent = ({size, type, children}) => {
  let variant = size && type ? `${size}-${type}` : size || type
  return <button className={props.variant}>{props.children}</button>
}

return <ButtonComponent size="large" type="danger">Delete project!</ButtonComponent>
```

To put it another way, make the API into your component match the API into your styles.

Does that make sense?

**Comment by nkbt** [comment]
@geelen I reckon you've meant:

```js
let ButtonComponent = ({variant, children} => (
  <button className={styles[variant]}>{props.children}
</button>)
```

**Comment by geelen** [comment]
Quite right!

**Comment by kgoggin** [comment]
Awesome, thanks so much everyone. In our context (large SaaS app with white-label branding) I can never really know for sure that any particular combination will or won't get used so I need to account for anything. That said, I really do appreciate the "constraint" of defining every combo up front. We're making a decision of what CSS methodology/framework to adopt and I'll definitely be pointing my team to this thread. Really appreciate the guidance!

**Comment by geelen** [comment]
One more thing, if you do want the multiple-class options and you're using React, https://github.com/gajus/react-css-modules might be what you're after.

**Comment by CaseReed** [comment]
> @geelen I reckon you've meant:
> 
> ```js
> let ButtonComponent = ({variant, children} => (
>   <button className={styles[variant]}>{props.children}
> </button>)
> ```

Amazing 🤘🔥
